bodhi/tests/server/test_models.py

```diff
@@ -216,6 +216,106 @@ def test_url(self):
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        self.update.builds.append(build)
+
+    def test_same_build_types(self):
+        """Assert when all builds are the same type, validation passes."""
+        build1 = model.RpmBuild(
+            nvr='the-greatest-package-1.0.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build2 = model.RpmBuild(
+            nvr='the-greatest-package-1.1.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        self.update.builds += [build1, build2]
+
+    def test_different_build_types(self):
+        """Assert when all builds are a different type, validation passes."""
+        build1 = model.Build(
+            nvr='the-greatest-package-1.0.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build2 = model.RpmBuild(
+            nvr='the-greatest-package-1.1.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        self.update.builds.append(build1)
+        self.assertRaises(ValueError, self.update.builds.append, build2)
+
+    def test_backref_no_builds(self):
+        """Assert when the first build is appended via a backref, the validator passes."""
+        build = model.Build(
+            nvr='the-greatest-package-1.0.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build.update = self.update
+
+    def test_backref_same_build_types(self):
+        """Assert when all builds are the same type and set via backref validation passes."""
+        build1 = model.RpmBuild(
+            nvr='the-greatest-package-1.0.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build2 = model.RpmBuild(
+            nvr='the-greatest-package-1.1.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build1.update = self.update
+        build2.update = self.update
+
+    def test_backref_different_build_types(self):
+        """Assert when builds differ in type and are set via backref validation passes."""
+        build1 = model.Build(
+            nvr='the-greatest-package-1.0.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build2 = model.RpmBuild(
+            nvr='the-greatest-package-1.1.0-fc17.1',
+            package_id=self.package.id,
+            release_id=self.update.release.id,
+        )
+        build1.update = self.update
+        with self.assertRaises(ValueError) as cm:
+            build2.update = self.update
+
+        self.assertEqual(str(cm.exception), u'An update must contain builds of the same type.')
```

Comment:
When looking at the code I was allllmost going to add a comment asking if the backrefs still get validated, but then I decided to go see what tests you wrote and here it is. Bam.